specs/capella/beacon-chain.md

```diff
@@ -0,0 +1,163 @@
+
+## Constants
+
+## Preset
+
+### State list lengths
+
+| Name | Value | Unit | Duration |
+| - | - | :-: | :-: |
+| `WITHDRAWAL_RECEIPT_LIMIT` | `uint64(2**40)` (= 1,099,511,627,776) | withdrawal receipts|
+
+## Configuration
+
+## Containers
+
+### Extended Containers
+
+#### `BeaconState`
+
+```python
+class BeaconState(Container):
+    # Versioning
+    genesis_time: uint64
+    genesis_validators_root: Root
+    slot: Slot
+    fork: Fork
+    # History
+    latest_block_header: BeaconBlockHeader
+    block_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
+    state_roots: Vector[Root, SLOTS_PER_HISTORICAL_ROOT]
+    historical_roots: List[Root, HISTORICAL_ROOTS_LIMIT]
+    # Eth1
+    eth1_data: Eth1Data
+    eth1_data_votes: List[Eth1Data, EPOCHS_PER_ETH1_VOTING_PERIOD * SLOTS_PER_EPOCH]
+    eth1_deposit_index: uint64
+    # Registry
+    validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
+    balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
+    # Randomness
+    randao_mixes: Vector[Bytes32, EPOCHS_PER_HISTORICAL_VECTOR]
+    # Slashings
+    slashings: Vector[Gwei, EPOCHS_PER_SLASHINGS_VECTOR]  # Per-epoch sums of slashed effective balances
+    # Participation
+    previous_epoch_participation: List[ParticipationFlags, VALIDATOR_REGISTRY_LIMIT]
+    current_epoch_participation: List[ParticipationFlags, VALIDATOR_REGISTRY_LIMIT]
+    # Finality
+    justification_bits: Bitvector[JUSTIFICATION_BITS_LENGTH]  # Bit set for every recent justified epoch
+    previous_justified_checkpoint: Checkpoint
+    current_justified_checkpoint: Checkpoint
+    finalized_checkpoint: Checkpoint
+    # Inactivity
+    inactivity_scores: List[uint64, VALIDATOR_REGISTRY_LIMIT]
+    # Sync
+    current_sync_committee: SyncCommittee
+    next_sync_committee: SyncCommittee
+    # Execution
+    latest_execution_payload_header: ExecutionPayloadHeader
+    # Withdrawals
+    withdrawal_receipts: List[WithdrawalReceipt, WITHDRAWAL_RECEIPT_LIMIT]  # [New in Cappela]
+```
+
+### New containers
+
+#### `WithdrawalReceipt`
+
+```python
+class WithdrawalReceipt(Container):
+    index: WithdrawalReceiptIndex
+    address: ExecutionAddress
+    amount: Gwei
+```
+
+## Helpers
+
+### Beacon state mutators
+
+#### `withdraw`
+
+```python
+def withdraw(state: BeaconState, index: ValidatorIndex, amount: Gwei) -> None:
+    # Decrease the validator's balance
+    decrease_balance(state, index, amount)
+    # Create a corresponding withdrawal receipt
+    receipt = WithdrawalReceipt(
+        index=WithdrawalReceiptIndex(len(state.withdrawal_receipts)),
+        address=state.validators[index].withdrawal_credentials[12:],
+        amount=amount,
+    )
+    state.withdrawal_receipts.append(receipt)
+```
+
+### Predicates
+
+#### `is_withdrawable_validator`
+
+```python
+def is_withdrawable_validator(validator: Validator, epoch: Epoch) -> bool:
+    """
+    Check if ``validator`` is withdrawable.
+    """
+    return validator.withdrawable_epoch <= epoch
```

Comment:
```suggestion
    return validator.withdrawable_epoch >= epoch
```

Reply:
I believe this should be `<=`. If the withdrawable epoch is at or in the past (less than) the `epoch` in question, then the validator is withdrawable.

Look at the `is_active_validator` logic for a parallel

Reply:
Take this extreme as example -- if withdrawable_epoch is `FAR_FUTURE_EPOCH`, then they will never be withdrawable. but with your suggestion, the > sign would def make them withdrawable

Reply:
You're right. This is a bit counter-intuitive to read hence my confusion.